### PR TITLE
Implement `max_keep_alive_requests` setting for server

### PR DESCRIPTION
**Pull request by nickitat**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
ClickHouse server now supports new setting `max_keep_alive_requests`. For keep-alive HTTP connections to the server it works in tandem with `keep_alive_timeout` - if idle timeout not expired but there already more than `max_keep_alive_requests` requests done through the given connection - it will be closed by the server.

---

We multiple times observed a problem when users created significantly imbalanced load on cluster nodes using keep-alive connections that lived for very long time. Some of those users implemented similar logic on their side and for some we just end up disabling KA on server side. This PR is aimed to solve the problem once and forever.

[Discussion](https://clickhouse-inc.slack.com/archives/C02F2LML5UG/p1703080006026579?thread_ts=1702934440.604579&cid=C02F2LML5UG) in CH private slack.

---

All the essential changes are:
* in programs/server/Server.cpp
* in src/IO/HTTPCommon.cpp

to have access to keep alive settings I exposed them through the `Session` object that is in turn exposed through `HTTPServerResponse`. the rest of diff is just changes is signatures of some methods.

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [x]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit c153fae0b8377aeca5e636a7ad0370e6ada42688 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/61793/c153fae0b8377aeca5e636a7ad0370e6ada42688/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Core/ServerSettings.h [hunk 1]
```diff
@@ -113,6 +113,7 @@ namespace DB
     M(Bool, async_load_databases, false, "Enable asynchronous loading of databases and tables to speedup server startup. Queries to not yet loaded entity will be blocked until load is finished.", 0) \
     M(Bool, display_secrets_in_show_and_select, false, "Allow showing secrets in SHOW and SELECT queries via a format setting and a grant", 0) \
     M(Seconds, keep_alive_timeout, DEFAULT_HTTP_KEEP_ALIVE_TIMEOUT, "The number of seconds that ClickHouse waits for incoming requests before closing the connection.", 0) \
+    M(UInt64, max_keep_alive_requests, 10000, "The number of seconds that ClickHouse waits for incoming requests before closing the connection.", 0) \
```
Comment (mlex) [inline src/Core/ServerSettings.h:116]: Copy paste bug. 

```suggestion
    M(UInt64, max_keep_alive_requests, 10000, "The maximum number of requests handled via a single http keepalive connection before the server closes this connection.", 0) \
```

**Review by serxa — APPROVED** [review]
LGTM. Let's also add a commented setting into `config.xml` with a small description

**Comment by CheSema** [comment]
Did you considered to implement this approach to close connection gracefully?
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408

HINT: There is a race condition between client and server. Client could still try to use already closed connection. It results in ugly error "No message received" or similar and retries. To avoid that case server could send 408 even to the idle connection right before closing it.

**Comment by CheSema** [comment]
Also I would be glad to see here some metrics which just shows does it work properly or nor.
Like current numbers persisted connection, number closed and timeouts. It could be useful in investigation if any.

**Comment by CheSema** [comment]
And one more thing, maybe some command `reset all persistent connection` also could be useful. In some cases of disbalance persistent connection, it could be used an ad-hoc mitigation.

File: src/IO/HTTPCommon.cpp [hunk 2]
```diff
@@ -33,14 +33,20 @@ namespace ErrorCodes
...
-void setResponseDefaultHeaders(HTTPServerResponse & response, size_t keep_alive_timeout)
+void setResponseDefaultHeaders(HTTPServerResponse & response)
 {
     if (!response.getKeepAlive())
         return;
 
-    Poco::Timespan timeout(keep_alive_timeout, 0);
-    if (timeout.totalSeconds())
-        response.set("Keep-Alive", "timeout=" + std::to_string(timeout.totalSeconds()));
+    const size_t keep_alive_timeout = response.getSession().getKeepAliveTimeout();
```
Comment (CheSema) [inline src/IO/HTTPCommon.cpp:41]: Here we get ka timeout from server's config. I do not see where we are reading ka timeout from client's request.

Comment (nickitat) [reply src/IO/HTTPCommon.cpp:42]: afaiu it was addressed in your PR

**Review by serxa — APPROVED** [review]

**Pull request merged**